Add frontend caches to release workflow

kschiffer opened the pull request:
#### Summary
This quickfix PR adds caching for frontend builds to the snapshot github action workflow. This will speed up builds as well as provide caches for PR builds.

Closes #3480 

#### Changes
- Add public folder cache to `release-snapshot` workflow
- Add babel cache to `release-snapshot` workflow
- Simplify frontend steps in `release-snapshot` workflow
- Simplify cache key for public folder

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

kschiffer committed:
dev: Add frontend caches to release workflow

htdvisser commented on .github/workflows/e2e.yml:
```diff
@@ -90,7 +90,7 @@ jobs:
         uses: actions/cache@v2
         with:
           path: public
-          key: public-cache-${{ hashFiles('pkg/webui/**') }}-${{ hashFiles('sdk/js/**/*.js', 'sdk/js/generated/*.json') }}-${{ hashFiles('config/webpack.config.babel.js') }}-${{ hashFiles('yarn.lock', 'sdk/js/yarn.lock')}}
```
Would it make sense to move this step a little closer to where it's actually used? So above L145?

kschiffer replied:
Other than a cosmetic improvement it could save a couple seconds when an earlier step fails. :shrug:

htdvisser replied:
Yeah, it would indeed just be to keep related things together.

htdvisser commented on .github/workflows/e2e.yml:
```diff
@@ -15,6 +15,7 @@ on:
       - 'yarn.lock'
       - 'cypress/**'
       - 'docker-compose.yml'
+      - '.github/e2e.yml'
```
```suggestion
      - '.github/workflows/e2e.yml'
```

htdvisser commented on .github/workflows/release-snapshot.yml:
```diff
@@ -42,6 +42,20 @@ jobs:
+        key: ${{ runner.os }}-babel-cache-${{ hashFiles('config/babel.config.json', 'config/webpack.config.babel.js') }}
+        restore-keys: |
+          ${{ runner.os }}-babel-cache-
+    - name: Initialize public folder cache
```
Would it make sense to move this step a little closer to where it's actually used? So above `name: Build frontend`?

htdvisser reviewed (commented):
Would it make sense to also add this caching to the `release-tag` workflow?

kschiffer commented:
@htdvisser I've moved the caching steps and also introduced some simplifications for the frontend build steps of the release workflow. Also my editor suggested a different indentation, which I believe is correct. You should set GitHub to hide whitespace changes for the review.

htdvisser commented on .github/workflows/release-snapshot.yml:
```diff
@@ -13,105 +13,113 @@ jobs:
+          restore-keys: |
+            ${{ runner.os }}-babel-cache-
+      - name: Build frontend
+        run: tools/bin/mage js:build
```
What happens here if the cache doesn't have a direct hit, but is still able to restore another build? Wouldn't omitting the `clean` result in old files accumulating in the public folder?

kschiffer replied:
The public cache has no restore keys so it will only work with direct hits (in which case the `js:build` step is not run anyway). The babel cache will use the `.cache` folder which is picked up by babel when transpiling the modules. There will be no effects for the `public` folder. It is also necessary to omit the `js:clean` command to leverage the babel cache, which is otherwise deleted by that command.
